src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -302,6 +315,35 @@ private string GetAnalyzerPath(AnalyzerReference analyzerReference)
             return analyzerReference.FullPath;
         }
 
+        protected override void ApplyCompilationOptionsChanged(ProjectId projectId, CompilationOptions options)
+        {
+            if (projectId == null)
+            {
+                throw new ArgumentNullException(nameof(projectId));
+            }
+
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            GetProjectData(projectId, out var hostProject, out var hierarchy, out var project);
+            foreach (Configuration configuration in project.ConfigurationManager)
+            {
+                switch (hostProject.Language)
+                {
+                    case LanguageNames.CSharp:
+                        var csharpProperties = (VSLangProj80.CSharpProjectConfigurationProperties3)configuration.Object;
+
+                        csharpProperties.AllowUnsafeBlocks = ((CSharpCompilationOptions)options).AllowUnsafe;
```

Comment:
I'd be curious what this looks like if we use IVsBuildPropertyStorage instead. It means we don't have to be looping through configurations/platforms, and would also have VB support.

@Pilchie, @davkean: any concerns with trying IVsBuildPropertyStorage?

Reply:
@jasonmalinowski It works great! Thanks!

Reply:
VB support for unsafe?

Reply:
Well, almost great... The VS property page is not updated and the errors don't go away until you reload the project.

Reply:
It only takes affect after a reload.

Reply:
@jasonmalinowski I don't think I'm capable of fixing that unfortunately. I'd love it if this did the right thing and added this to non-configuration specific properties but at this point I think someone else needs take care of this who knows more about the integration with VS/the project system here. Do you think it would be acceptable to leave this as adding to specific configurations (fixing #23342) and create a followup issue?

Reply:
Tagging @Pilchie to have somebody help investigate. I think we have aways to then poke the project system to redo design time builds and cause it to refresh.

Frankly, had we remembered IVsBuildPropertyStorage was a thing when this was written, I doubt we ever would have checked in what we checked in.

Reply:
@Pilchie For .NET Framework projects, it fails to refresh until a project reload and for SDK style projects, it seem that IVsBuildPropertyStorage isn't even implemented (the cast fails). So this doesn't worth with any kind of project in an acceptable way at the moment :-/

Reply:
We implement - IVsBuildPropertyStorage, it's just very likely you are casting the fake hierarchy to it. @jasonmalinowski How do you get to the real hierarchy?

Reply:
However, in saying that you might have troubles having the language service get updated & respect this setting in legacy project system if you bypass setting. SDK-based projects will just pick it up, however.

Reply:
@jasonmalinowski @davkean Here's the code I'm talking about (I created a new branch):
https://github.com/Neme12/roslyn/blob/8bf20859bd719ca015f778f75744441f796d7103/src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs#L364-L367

Reply:
Interestingly, this code *does* work with SDK-based projects:
```c#
var solution = (IVsSolution)DeferredState.ServiceProvider.GetService(typeof(SVsSolution));          
solution.GetProjectOfUniqueName(TryGetDTEProject(projectId).UniqueName, out var hierarchy);
var storage = (IVsBuildPropertyStorage)hierarchy;
```
and it is updated accordingly.

Reply:
> and the project is updated accordingly (still the same behavior with "legacy" projects).
Hmm. Are you sure the same behavior as legacy? It won't be instant but it will be picked up next design-time build that is triggered on the modification.

Legacy will never pick up the changes made by IVsBuildPropertyStorage.

Reply:
@davkean No, not same behavior as legacy, it works perfectly. But why do I have to get the hierarchy this way and the usual GetHierarchy doesn't work? What sort of "fake hierarchy" is that?

You're right, with legacy the changes aren't picked up, is there really no way to work around that, somehow forcing it to refresh? I don't think this behavior is acceptable.

Reply:
GetHierarchy dance is complicated and happens because we support multiple "roslyn projects" from a single project, whereas legacy only supports a single one. This will cleaned up one of these days.

There's no way to make legacy pick up the changes - legacy doesn't update it settings based on the project file unlike SDK-projects. It only writes to it.

Reply:
@jasonmalinowski What to do then? Should we have 2 different implementations for each project type?

Reply:
@davkean @tmeschter I thought there was a way to trigger a design time build which would update legacy (at the cost of a long, blocking operation)?

Reply:
For the legacy case, if we just update our workspace model, will the next design time build (whenever it is) pick up the changes that were made? Traditionally, whenever we've done TryApplyChanges we call back to the project system, and expect it to call back with us to update our state. For example, we don't add a file directly, we call DTE to add the file, and we get the new file notification back from the project system and handle it that way. But we don't *have* to do that and could update the workspace at the same time, and could possibly avoid the perf hit?

What's the cost of using those DTE configuration stuff today? I could imagine editing already triggers a new design-time build, so even giving the project system a kick might not be worse than today.

> @jasonmalinowski How do you get to the real hierarchy?

I was about to ask you. :smile: We normally never cared about it, I think, we usually are handling the stub hierarchy you gave us on project construction. What is the proper way to jump back? I'm not sure this is something we do today, actually.

Reply:
note: If I go to the project properties page, it is clearly marked as dirty, but the change isn't reflected the UI
![image](https://user-images.githubusercontent.com/11444821/38392709-c6a8eafa-3928-11e8-8773-ee087b27fd5d.png)

I don't know more about design-time builds and when they happen... I tried adding/removing references and it does indeed refresh intellisense and fix the errors, but the properties page is still not updated.

Reply:
Is there anything I can do here? I also tried doing this using Microsoft.Build APIs with no success.

Reply:
@jasonmalinowski I stumbled upon VSProject.Refresh which seemed like it fixed the issue because it refreshed the language service and gets rid of the errors. I assume this method triggers the design-time build. But this still has the problem that the property page in VS is not getting updated, and after more investigation, I also found out that this really only refreshed the current configuration - if I switch to another one, the error is back. I tried many other things and various hacks and don't think I can come up with a working solution here except for having is 2 separate implementations - using IVsBuildPropertyStorage for CPS projects but still using the old approach with "legacy" projects. The fact that VS itself can only insert these properties for each configuration separately with the UI even when selecting "All Configurations" doesn't give me much hope that I can fix this. So I hope what I did is fine. Having 2 implementations is definitely not good for maintainability and such, but I made sure I added tests for both.

Reply:
@Pilchie @Neme12 @jasonmalinowski With the older project system there isn't really a fully general way to say "something has changed so you better run a design-time build to update the language service". That project system basically expects to always sit as an intermediary between the MSBuild project and the language service, and to fully understand all the options that may be set.